### PR TITLE
Enable options_from_form(spawner, form_data) from configuration file

### PR DESCRIPTION
**Pull request by rcthomas**
In #3789 it was noted that attempting to customize `c.Spawner.options_from_form` would result in a syntax error if the callable signature was like `my_options_from_form(spawner, form_data)`.  Since the default `options_from_form` and subclass `options_from_form` implementations can take the spawner object and the form data as arguments, it appears to be a bug.

The question is how best to handle both cases, and here I document one that seems to work in my case and at least passes existing tests (though a new test with a configured `options_from_form` setting may make sense).  There might be a nicer way to do this than to check whether `options_from_form` is being a function (as opposed to a bound method).  If there are better ideas I'm happy to try them.

### CODE REVIEW

**Comment by minrk** [comment]
It's not _exactly_ a bug - the spec is that it should be a function taking only the form data, but it could make use of the spawner if it happens to be a bound method (where `self` is passed implicitly). But it does seem reasonable to _allow_ passing the spawner.

Rather than inspecting the _type_, which would break any existing pure functions `options_from_form` like:

```python
def options_from_form(options_form):
    return {
        'number': int(options_form['number'][0])
    }

c.Spawner.options_from_form = options_from_form
```

I think it makes sense to inspect the signature with `inspect.signature`. If it takes an additional `spawner` keyword argument, pass that along:

```python
sig = inspect.signature(options_from_form)
if 'spawner' in sig.parameters:
    options_from_form(options_form, spawner=self)
else:
    options_from_form(options_form)
```

That ought to be a backward-compatible way to introduce the feature.

**Comment by rcthomas** [comment]
Right, it's not _exactly_ a bug.  I'll rework to use `inspect` here.

File: jupyterhub/spawner.py [hunk 1]
```diff
@@ -424,6 +425,13 @@ def _options_from_form(self):
     def _default_options_from_form(self, form_data):
         return form_data
 
+    def run_options_from_form(self, form_data):
+        sig = signature(self.options_from_form)
+        if 'spawner' in sig.parameters:
+            return self.options_from_form(form_data, spawner=self)
+        else:
+            return self.options_from_form(form_data)
```
Comment (minrk) [inline jupyterhub/spawner.py:433]: Perfect! Can you add a test to test_spawner.py that exercises both forms of `options_from_form`? I was a little surprised the previous form of this PR didn't make anything fail, which means we didn't have coverage.

Sample form data is always a dict of lists of strings:

```python
from_data = {"key": ["value"]}
```

Comment (rcthomas) [reply jupyterhub/spawner.py:433]: I added 2 tests, one for a case with a spawner argument and one without.  Does that look OK?  Happy to improve it if it needs to do more.

**Comment by minrk** [comment]
Perfect, thanks!

**Pull request merged**